README.md
# @leichtgewicht/dns-packet
[![](https://img.shields.io/npm/v/@leichtgewicht/dns-packet.svg?style=flat)](https://www.npmjs.org/package/@leichtgewicht/dns-packet) [![](https://img.shields.io/npm/dm/@leichtgewicht/dns-packet.svg)](https://www.npmjs.org/package/@leichtgewicht/dns-packet) [![Tests](https://github.com/martinheidegger/dns-packet/actions/workflows/test.yml/badge.svg)](https://github.com/martinheidegger/dns-packet/actions/workflows/test.yml)

An [abstract-encoding](https://github.com/mafintosh/abstract-encoding) compliant module for encoding / decoding DNS packets. Lifted out of [multicast-dns](https://github.com/mafintosh/multicast-dns) as a separate module.

**This is a fork** of [dns-packet][] that comes with typescript types, javascript modules and does
not use Node.js primitives. Works well in browsers/react-native.

[dns-packet]: https://github.com/mafintosh/dns-packet

```
npm install @leichtgewicht/dns-packet
```

## UDP Usage

``` js
import * as dnsPacket from '@leichtgewicht/dns-packet'
import dgram from 'dgram'

const socket = dgram.createSocket('udp4')

const buf = dnsPacket.encode({
  type: 'query',
  id: 1,
  flags: dnsPacket.RECURSION_DESIRED,
  questions: [{
    type: 'A',
    name: 'google.com'
  }]
})

socket.on('message', message => {
  console.log(dnsPacket.decode(message)) // prints out a response from google dns
})

socket.send(buf, 0, buf.length, 53, '8.8.8.8')
```

Also see [the UDP example](examples/udp.js).

## TCP, TLS, HTTPS

While DNS has traditionally been used over a datagram transport, it is increasingly being carried over TCP for larger responses commonly including DNSSEC responses and TLS or HTTPS for enhanced security. See below examples on how to use `dns-packet` to wrap DNS packets in these protocols:

- [TCP](examples/tcp.js)
- [DNS over TLS](examples/tls.js)
- [DNS over HTTPS](examples/doh.js)

## API

#### `var buf = packets.encode(packet, [buf], [offset])`

Encodes a DNS packet into a buffer containing a UDP payload.

#### `var packet = packets.decode(buf, [offset])`

Decode a DNS packet from a buffer containing a UDP payload.

#### `var buf = packets.streamEncode(packet, [buf], [offset])`

Encodes a DNS packet into a buffer containing a TCP payload.

#### `var packet = packets.streamDecode(buf, [offset])`

Decode a DNS packet from a buffer containing a TCP payload.

#### `var len = packets.encodingLength(packet)`

Returns how many bytes are needed to encode the DNS packet

## Packets

Packets look like this

``` js
{
  type: 'query|response',
  id: optionalIdNumber,
  flags: optionalBitFlags,
  questions: [...],
  answers: [...],
  additionals: [...],
  authorities: [...]
}
```

The bit flags available are

``` js
import {
  RECURSION_DESIRED,
  RECURSION_AVAILABLE,
  TRUNCATED_RESPONSE,
  AUTHORITATIVE_ANSWER,
  AUTHENTIC_DATA,
  CHECKING_DISABLED
} from '@leichtgewicht/dns-packet'
```

To use more than one flag bitwise-or them together

``` js
var flags = packet.RECURSION_DESIRED | packet.RECURSION_AVAILABLE
```

And to check for a flag use bitwise-and

``` js
var isRecursive = message.flags & packet.RECURSION_DESIRED
```

A question looks like this

``` js
{
  type: 'A', // or SRV, AAAA, etc
  class: 'IN', // one of IN, CS, CH, HS, ANY. Default: IN
  name: 'google.com' // which record are you looking for
}
```

And an answer, additional, or authority looks like this

``` js
{
  type: 'A', // or SRV, AAAA, etc
  class: 'IN', // one of IN, CS, CH, HS
  name: 'google.com', // which name is this record for
  ttl: optionalTimeToLiveInSeconds,
  (record specific data, see below)
}
```

## Supported record types

#### `A`

``` js
{
  data: 'IPv4 address' // fx 127.0.0.1
}
```

#### `AAAA`

``` js
{
  data: 'IPv6 address' // fx fe80::1
}
```

#### `CAA`

``` js
{
  flags: 128, // octet
  tag: 'issue|issuewild|iodef',
  value: 'ca.example.net',
  issuerCritical: false
}
```

#### `CNAME`

``` js
{
  data: 'cname.to.another.record'
}
```

#### `DNAME`

``` js
{
  data: 'dname.to.another.record'
}
```

#### `DNSKEY`

``` js
{
  flags: 257, // 16 bits
  algorithm: 1, // octet
  key: Buffer
}
```

#### `DS`

``` js
{
  keyTag: 12345,
  algorithm: 8,
  digestType: 1,
  digest: Buffer
}
```

#### `HINFO`

``` js
{
  data: {
    cpu: 'cpu info',
    os: 'os info'
  }
}
```

#### `MX`

``` js
{
  preference: 10,
  exchange: 'mail.example.net'
}
```

#### `NS`

``` js
{
  data: nameServer
}
```

#### `NSEC`

``` js
{
  nextDomain: 'a.domain',
  rrtypes: ['A', 'TXT', 'RRSIG']
}
```

#### `NSEC3`

``` js
{
  algorithm: 1,
  flags: 0,
  iterations: 2,
  salt: Buffer,
  nextDomain: Buffer, // Hashed per RFC5155
  rrtypes: ['A', 'TXT', 'RRSIG']
}
```

#### `NULL`

``` js
{
  data: Buffer('any binary data')
}
```

#### `OPT`

[EDNS0](https://tools.ietf.org/html/rfc6891) options.

``` js
{
  type: 'OPT',
  name: '.',
  udpPayloadSize: 4096,
  flags: packet.DNSSEC_OK,
  options: [{
    // pass in any code/data for generic EDNS0 options
    code: 12,
    data: Buffer.alloc(31)
  }, {
    // Several EDNS0 options have enhanced support
    code: 'PADDING',
    length: 31,
  }, {
    code: 'CLIENT_SUBNET',
    family: 2, // 1 for IPv4, 2 for IPv6
    sourcePrefixLength: 64, // used to truncate IP address
    scopePrefixLength: 0,
    ip: 'fe80::',
  }, {
    code: 'TCP_KEEPALIVE',
    timeout: 150 // increments of 100ms.  This means 15s.
  }, {
    code: 'KEY_TAG',
    tags: [1, 2, 3],
  }]
}
```

The options `PADDING`, `CLIENT_SUBNET`, `TCP_KEEPALIVE` and `KEY_TAG` support enhanced de/encoding. See [optionscodes.js](https://github.com/mafintosh/dns-packet/blob/master/optioncodes.js) for all supported option codes. If the `data` property is present on a option, it takes precedence. On decoding, `data` will always be defined.

#### `PTR`

``` js
{
  data: 'points.to.another.record'
}
```

#### `RP`

``` js
{
  mbox: 'admin.example.com',
  txt: 'txt.example.com'
}
```

#### `RRSIG`

``` js
{
  typeCovered: 'A',
  algorithm: 8,
  labels: 1,
  originalTTL: 3600,
  expiration: timestamp,
  inception: timestamp,
  keyTag: 12345,
  signersName: 'a.name',
  signature: Buffer
}
```

#### `SOA`

``` js
{
  data:
    {
      mname: domainName,
      rname: mailbox,
      serial: zoneSerial,
      refresh: refreshInterval,
      retry: retryInterval,
      expire: expireInterval,
      minimum: minimumTTL
    }
}
```

#### `SRV`

``` js
{
  data: {
    port: servicePort,
    target: serviceHostName,
    priority: optionalServicePriority,
    weight: optionalServiceWeight
  }
}
```

#### `TXT`

``` js
{
  data: 'text' || Buffer || [ Buffer || 'text' ]
}
```

When encoding, scalar values are converted to an array and strings are converted to UTF-8 encoded Buffers. When decoding, the return value will always be an array of Buffer.

If you need another record type, open an issue and we'll try to add it.

## License

MIT


CHANGELOG.md
# Version 5.4.0 - 2022-06-27

- Feature: Adding dedicated packet codec
- Feature: Adding dedicated query/response codec as multiple queries are only theoretically supported.

# Version 5.3.0 - 2022-06-22

- Refactor: ESM module support

# Version 5.2.0 - 2019-02-21

- Feature: Added support for de/encoding certain OPT options.

# Version 5.1.0 - 2019-01-22

- Feature: Added support for the RP record type.

# Version 5.0.0 - 2018-06-01

- Breaking: Node.js 6.0.0 or greater is now required.
- Feature: Added support for DNSSEC record types.

# Version 4.1.0 - 2018-02-11

- Feature: Added support for the MX record type.

# Version 4.0.0 - 2018-02-04

- Feature: Added `streamEncode` and `streamDecode` methods for encoding TCP packets.
- Breaking: Changed the decoded value of TXT records to an array of Buffers. This is to accomodate DNS-SD records which rely on the individual strings record being separated.
- Breaking: Renamed the `flag_trunc` and `flag_auth` to `flag_tc` and `flag_aa` to match the names of these in the dns standards.

# Version 3.0.0 - 2018-01-12

- Breaking: The `class` option has been changed from integer to string.

# Version 2.0.0 - 2018-01-11

- Breaking: Converted module to ES2015, now requires Node.js 4.0 or greater


types/index.d.ts
import { RecordType } from '../types.js';
import { RecordClass } from '../rcodes.js';
import { OptionCodes } from '../optioncodes.js';
import { OPCode } from '../opcodes.js';

export { RecordType } from '../types.js';
export { RecordClass } from '../rcodes.js';
export { OptionCodes } from '../optioncodes.js';
export { OPCode } from '../opcodes.js';

export interface Codec <Type> {
  encode(package: Type, buf?: Uint8Array, offset?: number): Uint8Array;
  decode(buf: Uint8Array, offset?: number): Type;
  encodingLength(packet: Type): number;
}

export interface Question {
 type: RecordType;
 name: string;
 class?: RecordClass | undefined;
}

export interface SrvData {
 target: string;
 port?: number;
 priority?: number | undefined;
 weight?: number | undefined;
}

export interface HInfoData {
 cpu: string;
 os: string;
}

export interface SoaData {
 mname: string;
 rname: string;
 serial?: number | undefined;
 refresh?: number | undefined;
 retry?: number | undefined;
 expire?: number | undefined;
 minimum?: number | undefined;
}

export type TxtData = string | Uint8Array | Array<string | Uint8Array>;

export interface CaaData {
 issuerCritical?: boolean | undefined;
 flags?: number | undefined;
 tag: string;
 value: string;
}

export interface MxData {
 preference?: number | undefined;
 exchange: string;
}

export interface BaseAnswer<T, D> {
 type: T;
 name: string;
 ttl?: number | undefined;
 class?: RecordClass | undefined;
 data: D;
}

/**
 * Record types for which the library will provide a string in the data field.
 */
export type StringRecordType = "A" | "AAAA" | "CNAME" | "DNAME" | "NS" | "PTR";

/**
 * Record types for which the library does not attempt to process the data
 * field.
 */
export type OtherRecordType =
 | "AFSDB"
 | "APL"
 | "AXFR"
 | "CDNSKEY"
 | "CDS"
 | "CERT"
 | "DHCID"
 | "DLV"
 | "HIP"
 | "IXFR"
 | "IPSECKEY"
 | "KEY"
 | "KX"
 | "LOC"
 | "NAPTR"
 | "NSEC3PARAM"
 | "SIG"
 | "SSHFP"
 | "TA"
 | "TKEY"
 | "TLSA"
 | "TSIG"
 | "URI";

export type StringAnswer = BaseAnswer<StringRecordType, string>;
export type SrvAnswer = BaseAnswer<"SRV", SrvData>;
export type HInfoAnswer = BaseAnswer<"HINFO", HInfoData>;
export type SoaAnswer = BaseAnswer<"SOA", SoaData>;
export type TxtAnswer = BaseAnswer<"TXT", TxtData>;
export type CaaAnswer = BaseAnswer<"CAA", CaaData>;
export type MxAnswer = BaseAnswer<"MX", MxData>;
export type NullAnswer = BaseAnswer<"NULL", Uint8Array>;
export type OptAnswer = BaseAnswer<"OPT", OptionData[]>;
export type DNSKeyAnswer = BaseAnswer<"DNSKEY", DNSKeyData>;
export type RRSigAnswer = BaseAnswer<"RRSIG", RRSigData>;
export type RPAnswer = BaseAnswer<"RP", RPData>;
export type NSecAnswer = BaseAnswer<"NSEC", NSecData>;
export type NSec3Answer = BaseAnswer<"NSEC3", NSec3Data>;
export type DSAnswer = BaseAnswer<"DS", DigestData>;
export type BufferAnswer = BaseAnswer<OtherRecordType, Uint8Array>;

export type Answer =
 | StringAnswer
 | SrvAnswer
 | HInfoAnswer
 | SoaAnswer
 | TxtAnswer
 | CaaAnswer
 | MxAnswer
 | NullAnswer
 | OptAnswer
 | DNSKeyAnswer
 | RRSigAnswer
 | RPAnswer
 | NSecAnswer
 | NSec3Answer
 | DSAnswer
 | BufferAnswer;

export interface Packet {
  /**
   * Whether the packet is a query or a response. This field may be
   * omitted if it is clear from the context of usage what type of packet
   * it is.
   */
  type?: "query" | "response" | undefined;

  id?: number | undefined;

  /**
   * A bit-mask combination of zero or more of:
   * {@link AUTHORITATIVE_ANSWER},
   * {@link TRUNCATED_RESPONSE},
   * {@link RECURSION_DESIRED},
   * {@link RECURSION_AVAILABLE},
   * {@link AUTHENTIC_DATA},
   * {@link CHECKING_DISABLED}.
   */
  flags?: number | undefined;
  questions?: Question[] | undefined;
  answers?: Answer[] | undefined;
  additionals?: Answer[] | undefined;
  authorities?: Answer[] | undefined;
  flag_qr?: boolean;
  opcode?: OPCode;
  flag_aa?: boolean;
  flag_tc?: boolean;
  flag_rd?: boolean;
  flag_ra?: boolean;
  flag_z?: boolean;
  flag_ad?: boolean;
  flag_cd?: boolean;
  rcode?: RecordClass;
}

export const DNSSEC_OK: 32768;
export const AUTHORITATIVE_ANSWER: 1024;
export const TRUNCATED_RESPONSE: 512;
export const RECURSION_DESIRED: 256;
export const RECURSION_AVAILABLE: 128;
export const AUTHENTIC_DATA: 32;
export const CHECKING_DISABLED: 16;

export interface DNSKeyData {
  key: string;
  flags: number;
  algorithm: number;
}
export interface DigestData {
  digest: Uint8Array;
  keyTag: number;
  algorithm: number;
  digestType: number;
}

export interface NSecData {
  nextDomain: Uint8Array;
  rrtypes: RecordType[];
}
export interface NSec3Data {
  salt: Uint8Array;
  nextDomain: Uint8Array;
  algorithm: number;
  flags: number;
  iterations: number;
  rrtypes: RecordType[];
}

export type ACodec = Codec<string>;
export type PtrCodec = Codec<Uint8Array>;
export type TxtCodec = Codec<Uint8Array[]>;
export type NullCodec = Codec<Uint8Array>;
export type CaaCodec = Codec<string> & {
  ISSUER_CRITICAL: 128
};
export type DNSKeyCodec = Codec<DNSKeyData> & {
  PROTOCOL_DNSSEC: 3
  ZONE_KEY: 0x80
  SECURE_ENTRYPOINT: 0x8000
};
export type DSCodec = Codec<DigestData>;
export type HInfoCodec = Codec<HInfoData>;
export type AAAACodec = Codec<string>;
export type UnknownCodec = Codec<Uint8Array>;
export type SrvCodec = Codec<SrvData>;
export type NSCodec = Codec<string>;
export type SoaCodec = Codec<SoaData>;
export type MxCodec = Codec<MxData>;
export type OptCodec = Codec<OptionData[]>;
export type QuestionCodec = Codec<Question>;
export type RRSigCodec = Codec<RRSigData>;
export type RPCodec = Codec<RPData>;
export type NSecCodec = Codec<NSecData>;
export type NSec3Codec = Codec<NSec3Data>;
export type PacketCodec = Codec<Packet>;

export type SingleQuestionPacket = Omit<Packet, 'questions'> & {
  question: Question
};

export type QueryCodec = Codec<SingleQuestionPacket>;
export type ResponseCodec = Codec<SingleQuestionPacket>;

export const a: ACodec;
export const caa: CaaCodec;
export const ptr: PtrCodec;
export const cname: PtrCodec;
export const dname: PtrCodec;
export const dnskey: DNSKeyCodec;
export const ds: DSCodec;
export const hinfo: HInfoCodec;
export const aaaa: AAAACodec;
export const answer: Codec<Answer>;
export const txt: TxtCodec;
export const unknown: UnknownCodec;
export const mx: MxCodec;
export const name: Codec<string>;
export const ns: NSCodec;
export const soa: SoaCodec;
export const srv: SrvCodec;
export const option: Codec<OptionData>;
export const opt: OptCodec;
export const question: QuestionCodec;
export const rrsig: RRSigCodec;
export const rp: RPCodec;
export const nsec: NSecCodec;
export const nsec3: NSec3Codec;
export const packet: PacketCodec;
export const query: QueryCodec;
export const response: ResponseCodec;

declare const rnull: NullCodec;
export { rnull as null };

export interface GenericOptionData {
  code: OptionCodes;
  data: Uint8Array;
}
export interface ClientSubnetOptionData {
  code: 'CLIENT_SUBNET';
  ip: string;
  family?: 1 | 2;
  sourcePrefixLength?: number;
  scopePrefixLength?: number;
}
export interface TCPKeepaliveOptionData {
  code: 'TCP_KEEPALIVE';
  timeout?: number;
}
export interface PaddingOptionData {
  code: 'PADDING';
  length?: number;
}
export interface KeyTagOptionData {
  code: 'KEY_TAG';
  tags: number[];
}

export type OptionData =
    GenericOptionData
  | ClientSubnetOptionData
  | TCPKeepaliveOptionData
  | PaddingOptionData
  | KeyTagOptionData;

export interface RRSigData {
  signature: Uint8Array;
  typeCovered: RecordType;
  algorithm: number;
  labels: number;
  originalTTL: number;
  expiration: number;
  inception: number;
  keyTag: number;
  signersName: string;
}

export interface RPData {
  mbox?: string;
  txt?: string;
}

export type AnyTypeCodec =
  ACodec | PtrCodec | TxtCodec | NullCodec | AAAACodec |
  SrvCodec | HInfoCodec | CaaCodec | NSCodec | SoaCodec |
  MxCodec | OptCodec | RRSigCodec | RPCodec | NSecCodec |
  NSec3Codec | DSCodec | UnknownCodec;

export type TypeCodec <Type> =
  Type extends string ?
    Uppercase<Type> extends 'A' ? ACodec :
    Uppercase<Type> extends 'PTR' ? PtrCodec :
    Uppercase<Type> extends 'CNAME' ? PtrCodec :
    Uppercase<Type> extends 'DNAME' ? PtrCodec :
    Uppercase<Type> extends 'TXT' ? TxtCodec :
    Uppercase<Type> extends 'NULL' ? NullCodec :
    Uppercase<Type> extends 'AAAA' ? AAAACodec :
    Uppercase<Type> extends 'SRV' ? SrvCodec :
    Uppercase<Type> extends 'HINFO' ? HInfoCodec :
    Uppercase<Type> extends 'CAA' ? CaaCodec :
    Uppercase<Type> extends 'NS' ? NSCodec :
    Uppercase<Type> extends 'SOA' ? SoaCodec :
    Uppercase<Type> extends 'MX' ? MxCodec :
    Uppercase<Type> extends 'OPT' ? OptCodec :
    Uppercase<Type> extends 'DNSKEY' ? DNSKeyCodec :
    Uppercase<Type> extends 'RRSIG' ? RRSigCodec :
    Uppercase<Type> extends 'RP' ? RPCodec :
    Uppercase<Type> extends 'NSEC' ? NSecCodec :
    Uppercase<Type> extends 'NSEC3' ? NSec3Codec :
    Uppercase<Type> extends 'DS' ? DSCodec :
    UnknownCodec
  : AnyTypeCodec;

export function enc <Type>(type: Type): TypeCodec<Type>;

export function encode(packet: Packet, buf?: Uint8Array, offset?: number): Uint8Array;
export function decode(buf: Uint8Array, offset?: number): Packet;
export function encodingLength(packet: Packet): number;
export function decodeList <T= any>(list: T[], codec: Codec<T>, buf: Uint8Array, offset?: number): Packet[];
export function encodeList <T= any>(list: T[], codec: Codec<T>, buf?: Uint8Array, offset?: number): number;
export function encodingLengthList <T= any>(list: T[], codec: Codec<T>): number;
export function streamDecode(buffer: Uint8Array): Packet | null;
export function streamEncode(packet: Packet): Uint8Array;
